find_email_addresses.py

```diff
@@ -0,0 +1,261 @@
+
+    extensions = ['pdf', 'jpg', 'png', 'gif', 'zip', 'doc', 'docx']
+
+    url = urllib.parse.urlparse(link)
+
+    for ext in extensions:
+        if url.path.lower().endswith(ext.lower()):
+            return True
+
+    return False
+
+def _get_links_from_string(string, domain, ignore_binary=True):
+    """Returns internal links found in a string.
+
+    Args:
+        s: The string to search
+        domain: The local domain name as a string
+
+    Returns:
+        A deque object containing unique internal links found in s."""
+
+    links = deque()
+    # This uses the RFC 3986 URI definition (see
+    # http://tools.ietf.org/html/rfc3986#section-2) and matches 'src="<URI>"'
+    # and 'href="<URI>"'.
+    link_regex = re.compile(r"""(?:(?:src|href)=["'])((?:(?:[a-z]+:)|"""
+                            r'(?:\.{1,2}))?\/{1,2}'
+                            r"[\w\-\.~:\/\?#\]\[@!\$&'\(\)\*\+,;=%]+)"
+                            r"""(?:["'])""", re.I)
+
+    for link in re.findall(link_regex, string):
+        if (_is_internal_link(link, domain) and not links.count(link)
+                and not (ignore_binary and _is_binary_link(link))):
+            links.append(link)
+
+    return links
+
+def _decode_bytestring(string):
+    """Decodes and returns a bytestring as a string
+
+    Args:
+        s: The bytestring to decode
+
+    Returns:
+        A decoded string or an empty string if encoding cannot be determined"""
+
+    enc = chardet.detect(string)
+
+    try:
+        if enc['encoding']:
+            return string.decode(enc['encoding'])
+    # Ignore pages that don't decode. These are usually low confidence anyway.
+    except UnicodeDecodeError:
+        pass
+
+    return ""
+
+def _assemble_url(link, domain, scheme):
+    """Generates a properly formatted URI
+
+    Args:
+        link: The raw link to format
+        domain: The domain to use if link does not specify
+        scheme: The scheme to use if link does not specify
+
+    Returns:
+        A properly formatted URL as a string"""
+
+    string = ''
+    url = urllib.parse.urlparse(link)
+
+    if url.scheme:
+        string += url.scheme + '://'
+    else:
+        string += scheme + '://'
+
+    if url.netloc:
+        string += url.netloc
+    else:
+        string += domain
+
+    string += url.path
+
+    if url.params:
+        string = string + ';' + url.params
+    if url.query:
+        string = string + '?' + url.query
+    if url.fragment:
+        string = string + '#' + url.fragment
+
+    return string
+
+def _is_valid_domain(domain):
+    """Determines if the provided domain is in a valid format.
+
+    Args:
+        domain: The string to test
+
+    Returns:
+        True if the domain is valid, false otherwise."""
+
+    # For some reason, urlparse won't recognize the domain without the scheme,
+    # so adding a scheme is a bit of a hack, but it works. It will even work
+    # if someone enters "http://domain.com" as the domain.
```

Comment:
Sorry, that comment wasn't terribly clear. What I meant was that since the challenge called for a domain to be entered (not a url), `http://jana.com` is not a valid input.

If the user enters `http://jana.com` for the domain, if you add the `http://` on top, you get:

```python
>>> urllib.parse.urlparse("http://http://jana.com")
ParseResult(scheme='http', netloc='http:', path='//jana.com', params='', query='', fragment='')
```

Both `netloc` and `path` are populated - not correctly, but it is consistent.

If you only enter `jana.com`, without adding the `http://`, you get:

```python
ParseResult(scheme='', netloc='', path='jana.com', params='', query='', fragment='')
```

I suppose, upon further investigation, that the best workaround for this would be to check if `path` is populated while `netloc` is an empty string, since that would work for subdomains and domains with paths:

```python
>>> urllib.parse.urlparse("jana.com/contact")
ParseResult(scheme='', netloc='', path='jana.com/contact', params='', query='', fragment='')
>>> urllib.parse.urlparse("blog.jana.com")
ParseResult(scheme='', netloc='', path='blog.jana.com', params='', query='', fragment='')
>>> urllib.parse.urlparse("blog.jana.com/contact")
ParseResult(scheme='', netloc='', path='blog.jana.com/contact', params='', query='', fragment='')
```